### PR TITLE
[clang-doc] fix bug introduced by asset test

### LINKED ISSUE #97507: Test clang-tools-extra/test/clang-doc/basic-project.test failing when built using non-ninja generator
We have an internal job which attempts to build the LLVM tree using the Visual Studio 2019 builder configuration (non-ninja) and it recently started hitting a test failure of clang-tools-extra/test/clang-doc/basic-project.test which I bisected back to when f14ad74478494cbd9d23af841d6f6b2808afda71 (#96358) was committed. The problem is that the change hard-codes how to find the dependent assets as `../share/clang-doc`. This is fine in ninja builds, but fails with the Visual Studio generator because the configuration name is part of the path.

For example, in a ninja build, the value of AssetsPath is (correctly) `C:\Dev\git\merge\build\bin\..\share\clang-doc\index.js`. But when using a Visual Studio generator, the value of AssetsPath is now `C:\Dev\git\merge\build\Debug\bin\..\share\clang-doc\index.js` which refers to a non-existent directory. In this particular case, it requires an extra `..` in the path to find the expected directory.

Lit test output:
```
FAIL: Clang Tools :: clang-doc/basic-project.test (1 of 1)
******************** TEST 'Clang Tools :: clang-doc/basic-project.test' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
rm -rf C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp && mkdir -p C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/docs C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/build
# executed command: rm -rf 'C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp'
# executed command: mkdir -p 'C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/docs' 'C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/build'
# RUN: at line 2
sed 's|$test_dir|C:/Dev/git/merge/clang-tools-extra/test/clang-doc|g' C:\Dev\git\merge\clang-tools-extra\test\clang-doc/Inputs/basic-project/database_template.json > C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/build/compile_commands.json
# executed command: sed 's|$test_dir|C:/Dev/git/merge/clang-tools-extra/test/clang-doc|g' 'C:\Dev\git\merge\clang-tools-extra\test\clang-doc/Inputs/basic-project/database_template.json'
# RUN: at line 3
clang-doc --format=html --output=C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/docs --executor=all-TUs C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/build/compile_commands.json
# executed command: clang-doc --format=html '--output=C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/docs' --executor=all-TUs 'C:\Dev\git\merge\build\tools\clang\tools\extra\test\clang-doc\Output\basic-project.test.tmp/build/compile_commands.json'
# .---command stdout------------
# | Emiting docs in html format.
# `-----------------------------
# .---command stderr------------
# | default index.js file missing at C:\Dev\git\merge\build\Debug\bin\..\share\clang-doc\index.js
# |
# `-----------------------------
# error: command failed with exit status: 1
```

### PR DESCRIPTION
**Pull request by PeterChou1**
Fixes https://github.com/llvm/llvm-project/issues/97507
this patch fixes the bug by copying assets to the correct repository for windows builds. It introduces a new global variable 
LLVM_SHARE_OUTPUT_INTDIR which points a platform agnostic shared directory

### CODE REVIEW

**Comment by ilovepi** [comment]
seems to fail on windows CI.

**Review by dyung — CHANGES_REQUESTED** [review]
I believe (and the pre-merge testing seems to validate my belief) that this breaks the case where you build with ninja + Visual Studio. Also, I believe the Xcode generator is similar to the Visual Studio generator in that they both are multi-config generators, so that configuration is likely failing as well.

**Comment by PeterChou1** [comment]
> I believe (and the pre-merge testing seems to validate my belief) that this breaks the case where you build with ninja + Visual Studio. Also, I believe the Xcode generator is similar to the Visual Studio generator in that they both are multi-config generators, so that configuration is likely failing as well.

@ilovepi 

I don't have access to a mac could you run build for checking clang-doc in xcode to see if it reports any errors?

File: clang-tools-extra/clang-doc/tool/CMakeLists.txt [hunk 1]
```diff
@@ -25,7 +25,11 @@ set(assets
 )
 
 set(asset_dir "${CMAKE_CURRENT_SOURCE_DIR}/../assets")
-set(resource_dir "${CMAKE_BINARY_DIR}/share/clang-doc")
+if(MSVC)
```
[the remark this answers is not in the document]
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/tool/CMakeLists.txt]: You're right this bug is specific to generators used I've update the code

**Comment by ilovepi** [comment]
> > I believe (and the pre-merge testing seems to validate my belief) that this breaks the case where you build with ninja + Visual Studio. Also, I believe the Xcode generator is similar to the Visual Studio generator in that they both are multi-config generators, so that configuration is likely failing as well.
> 
> @ilovepi
> 
> I don't have access to a mac could you run build for checking clang-doc in xcode to see if it reports any errors?

while I have access to a mac, I don't have the ability to use XCode due to company policy.

**Comment by petrhosek** [comment]
We already have set of CMake variables that we use to refer to `bin` or `lib` directory which account for various configurations: https://github.com/llvm/llvm-project/blob/c02e8f762a410e55581866c43636efcd6504c1bd/llvm/CMakeLists.txt#L446-L467

I think we should introduce another one for the `share` directory and use that.

**Comment by PeterChou1** [comment]
> We already have set of CMake variables that we use to refer to `bin` or `lib` directory which account for various configurations:
> 
> https://github.com/llvm/llvm-project/blob/c02e8f762a410e55581866c43636efcd6504c1bd/llvm/CMakeLists.txt#L446-L467
> 
> I think we should introduce another one for the `share` directory and use that.

I've introduced a variable in the root cmakelists.txt I'm not very familiar with cmake though I'm sure if this is the correct approach

**Comment by PeterChou1** [comment]
@dyung I do not have access to mac for xcode could you provide me with a log of the build failure on xcode

**Comment by dyung** [comment]
> @dyung I do not have access to mac for xcode could you provide me with a log of the build failure on xcode

```
% python3 ~/src/git/llvm-dyung/build/Debug/bin/llvm-lit -v basic-project.test
-- Testing: 1 tests, 1 workers --
FAIL: Clang Tools :: clang-doc/basic-project.test (1 of 1)
******************** TEST 'Clang Tools :: clang-doc/basic-project.test' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
Emiting docs in html format.

--
Command Output (stderr):
--
RUN: at line 1: rm -rf /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp && mkdir -p /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/docs /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/build
+ rm -rf /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp
+ mkdir -p /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/docs /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/build
RUN: at line 2: sed 's|$test_dir|/Users/dyung/src/git/llvm-dyung/clang-tools-extra/test/clang-doc|g' /Users/dyung/src/git/llvm-dyung/clang-tools-extra/test/clang-doc/Inputs/basic-project/database_template.json > /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/build/compile_commands.json
+ sed 's|$test_dir|/Users/dyung/src/git/llvm-dyung/clang-tools-extra/test/clang-doc|g' /Users/dyung/src/git/llvm-dyung/clang-tools-extra/test/clang-doc/Inputs/basic-project/database_template.json
RUN: at line 3: clang-doc --format=html --output=/Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/docs --executor=all-TUs /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/build/compile_commands.json
+ clang-doc --format=html --output=/Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/docs --executor=all-TUs /Users/dyung/src/git/llvm-dyung/build/tools/clang/tools/extra/test/clang-doc/Output/basic-project.test.tmp/build/compile_commands.json
default index.js file missing at /Users/dyung/src/git/llvm-dyung/build/Debug/bin/../share/clang-doc/index.js

```

**Comment by PeterChou1** [comment]
I've added the install path using a platform agnostic variable LLVM_RUNTIME_OUTPUT_INTDIR which should fix build path problems

**Review by ilovepi — APPROVED** [review]

**Comment by mgorny** [comment]
This change broke standalone clang builds against installed LLVM, since `LLVM_SHARE_OUTPUT_INTDIR` is not exported by LLVM and therefore resolves to an empty path:

```
ninja: error: mkdir(/clang-doc): Permission denied
ninja: build stopped: .
```

Why are you even declaring the directory inside LLVM when it's only used by Clang?

Please revert.

**Comment by mgorny** [comment]
Thank you.

**Pull request merged**